Prepare ground for v2

MMMalik opened the pull request:
Within this PR the following has been done:

1. Single React component has been split into `useCKEditor` hook and `CKEditor` component. The first one is supposed to contain core business logic and the latter is a convenience wrapper for simpler use cases. Advanced use cases will be left to userland via `useCKEditor` hook. Please note that business logic **is not completed yet**. It's just an early scaffold for future improvements (will be continued in #124).

2. TypeScript has been introduced. There is `tsconfig.json` file and TS has been added to build pipeline. Please notice that for the sake of unit testing `babel` is used to transpile TS code. It's due to some performance issues I had with `@rollup/plugin-typescript` inside Karma.

3. Rollup is used to create 3 output files, one for each format: `umd`, `esm`, & `cjs`. Additonally, TS declarations will be emitted. `package.json` has been adjusted to contain relevant fields (`main`, `module`, `types`).

4. Samples are contained in `samples` directory as self-contained applications. This approach will allow users to easily fork each sample (via GitHub or other services, e.g. CodeSandbox) for testing and issue reporting.

5. Approach to tests has been changed. Unit tests will run via Karma with the help of React Testing Library. Unit tests will run on all supported versions of React but _only_ on Chrome. The idea here is to just get a feedback if business logic is implemented correctly. In order to test if this library works correctly on all supported browsers, E2E tests are introduced. They are configured to run on all `samples` via Nightwatch.js and BrowserStack. Helper `scripts` have been adjusted accordingly. See my comment below for longer explanation.

Please read through `CONTRIBUTING.md` before reviewing.

CI pipeline might be still unstable though, especially E2E tests...

Closes #193.
Closes #184.
Closes #159.
Closes #124.

A sidenote on **testing**. I spent some time on choosing right way to test this library. Unit tests should be at the very core of any testing suite. In a React app, this is typically achieved with Jest framework running in `jsdom` environment on Node.js. However, this approach is not compatible with `ckeditor4` due to editor's custom dependency management system via injected `scripts`. Therefore, unit tests must run on a real browser. This has been achieved with custom Karma config. I believe **it is not necessary** to run unit tests on all supported browsers in case of this library. Anyway, since we still should test supported browsers somehow, I figured that it might be enough to run some sanity checks on `samples`. Therefore, I've added simple end-2-end tests with help of Nightwatch.js and BrowserStack (there is an [official guide](https://www.browserstack.com/docs/automate/selenium/getting-started/nodejs/nightwatch)). This approach has another benefit: We test library output files and not source code in opposition to unit tests. This way we can ensure that our library can be easily consumed and there are no issues with the build output.

Comandeer reviewed (changes requested):
Overally looks good! There are only some minor notes in several places and two bigger general doubts/questions:

* Shouldn't CJS and ESM builds also be minified?

* The `umd` sample can't be built. When I'm trying to run `npm install`, I get:

  ```
  npm ERR! code ERESOLVE
  npm ERR! ERESOLVE unable to resolve dependency tree
  npm ERR! 
  npm ERR! While resolving: undefined@undefined
  npm ERR! Found: react@17.0.2
  npm ERR! node_modules/react
  npm ERR!   react@"^17.0.2" from the root project
  npm ERR! 
  npm ERR! Could not resolve dependency:
  npm ERR! peer react@"^16.0.0" from ckeditor4-react@1.4.0
  npm ERR! node_modules/ckeditor4-react
  npm ERR!   ckeditor4-react@"latest" from the root project
  npm ERR! 
  npm ERR! Fix the upstream dependency conflict, or retry
  npm ERR! this command with --force, or --legacy-peer-deps
  npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
  ```

  Probably it's due to the mismatch between sample's deps and deps of the latest published version of `ckeditor4-react` package. It works when I use `--legacy-peer-deps` flag. Not sure if we shouldn't add info about it to the `CONTRIBUTING.md` file.

MMMalik commented on .github/workflows/check-pr.yml:
```diff
@@ -3,7 +3,7 @@ name: Check PR
 on:
     pull_request:
         branches:
-            - master
```
I've adjusted triggers in `check-pr` and `test-all` to run on `pull_request` to `master` and `major`. In subsequent tasks we might want to think how tests should be triggered and by what conditions. 

Related: #197

MMMalik commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,79 @@
+
+### Development
+
+First, make sure that `CHROME_BIN` environment variable is set, e.g. in your `.bashrc` file: `export CHROME_BIN=path_to_chrome_bin`. Then run the following command:
```
Right, `karma-chrome-launcher` makes some attempts to locate Chrome binary on your system. If it fails (or if you want to override the default), then use `CHROME_BIN` env. I've adjusted `CONTRIBUTING` to make it clearer.

MMMalik commented on samples/basic/src/App.jsx:
```diff
@@ -0,0 +1,13 @@
+function App() {
+	return (
+		<section>
+			<div title="React version">{`Running React v${ React.version }`}</div>
```
Yeah, that's an attempt to make this `div` easily discoverable by a simple `xpath` query. I've removed `title` in favor of a dummy `className` [here](https://github.com/ckeditor/ckeditor4-react/pull/199/files#diff-7a13a2ac88da6a82903eee46386c11a1d937e69c93798efcba3b9fc2d851ac2aR19).

Any other ideas? Maybe `id`?

Comandeer replied:
IMO both `[id]` and `[class]` are equally good solutions.

MMMalik commented on samples/umd/build.js:
```diff
@@ -0,0 +1,16 @@
+/* eslint-env node */
+
+const shell = require( 'shelljs' );
```
Good catch. Although I've used `shelljs` extensively in parent package, it's not necessary in this sample.

MMMalik commented on samples/umd/package.json:
```diff
@@ -0,0 +1,17 @@
+	"private": true,
+	"scripts": {
+		"build": "node ./build.js",
+		"start": "http-server"
```
Thanks, I've added `prestart` script.

MMMalik commented:
Thanks @Comandeer for the review.

Regarding your main concerns:

1. I don't see many use cases for minified versions of `esm` and `cjs` packages at this moment. Most likely a React app developer would use our library with the help of a bundler anyway, so I believe it's better to perform such transformations downstream. 
If at any point in the future community needs minified versions of these packages, then we can consider adding them (it won't be a breaking change).

2. Regarding failing build. Please note that `npm link ckeditor4-react` must be run in every sample separately and after every `npm install` in that sample (`install` will get rid of links). Try again please - it should be working fine. I've added some clarification to `CONTRIBUTING`.

MMMalik commented:
Hi @Comandeer,

I'm submitting `alpha` version for review. As discussed, I've merged core logic from #124 into this branch. To summarize, the following has been done:

1. Logic from `v1` has been rewritten with hooks.
2. Unit tests have been written from scratch with help of React Testing Library.
3. E2E tests have been added. Tests run on BrowserStack via Nighwatch.js. Only `samples/**` are tested this way.
4. There are multiple samples added (around 10 different ones).

From issues I've noticed so far:

1. Sometimes E2E tests fail (ca. 1% of test cases). These are false negatives resulting, as I suppose, from insufficient support for `iframe` in testing tools (Nightwatch? Selenium?). I've added a simple re-try mechanism to mitigate that.
2. I had to limit number of E2E tests run by narrowing tested React version to last two versions. Otherwise E2E tests for each browser take ca. 1h to complete. Alternatively we can run tests on all React versions but limit number of tested samples.

Comandeer commented on src/CKEditor.tsx:
```diff
@@ -0,0 +1,223 @@
+				editor,
+				evtName: 'instanceReady',
+				handler: ( { editor } ) => {
+					editor.setData( initData );
```
Maybe setting [`options.internal`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#method-setData) would solve the issues with undo (see overall comments about PR for more details) :thinking: 

```suggestion
					editor.setData( initData, {
						internal: true
					} );
```

MMMalik replied:
I've tried that but without luck...

The following `config` seems to be working fine though:

```js
{
	/**
	 * Prevents undo icon flickering.
	 */
	noSnapshot: true,

	/**
	 * Resets undo stack.
	 */
	callback: () => {
		editor.resetUndo();
	}
}
```

Comandeer commented on src/registerEditorEventHandler.ts:
```diff
@@ -0,0 +1,77 @@
+
+	let _handler: CKEditorEventHandler | undefined | null;
+
+	if ( handler ) {
```
Can't the function return early if `handler` is not provided?

MMMalik replied:
I've changed this function's interface a little so that `handler` must be passed. It doesn't really make sense to register a handler without, well, a handler ;-).

Comandeer reviewed (changes requested):
Overally looks good, but there are some weird issues:

* I'd slightly change how `test:e2e` script works. Currently it does not exit with code 1 when the initial connection to BS can't be made. It should exit early in this case, because any subsequent tests will fail.

* All E2E tests passed, but the whole process didn't end with 0 exit code. It failed with the following error:

  ```
  internal/process/per_thread.js:200
        throw errnoException(err, 'kill');
        ^
  
  Error: kill ESRCH
      at process.kill (internal/process/per_thread.js:200:13)
      at ChildProcess.<anonymous> (/Users/comandeer/repos/ckeditor4/ckeditor4-react/scripts/e2e-runner.js:163:13)
      at ChildProcess.emit (events.js:315:20)
      at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12) {
    errno: -3,
    code: 'ESRCH',
    syscall: 'kill'
  }
  ```

* Something's wrong with setting inital data. E.g. in `basic` example there are two undo steps in the editor from the start. Pressing undo button once shows "undefined" as the content of the editor and pressing it once more shows the empty editor. The `initialData` prop should not create any undo step at all. The same issue is present in other samples as well.

* In `component` sample the behaviour of undo is weird. It seems that the state is shared between inline and classic editor. E.g. switching from classic to inline one generates new editor, but with the undo step. Pressing undo button shows the data from the previous editor. There shouldn't be any undo step after switching editor's type. The same issue is present in the `hook` sample.

* Please update the list of used libs inside `LICENSE.md` file.

MMMalik commented on samples/README.md:
```diff
@@ -0,0 +1,18 @@
+| [re-order](re-order)                 | Example of handling multiple instances of `CKEditor` component (re-ordering case).              | Insert link |
+| [ssr](ssr)                           | Showcase of using `CKEditor` with React server-side rendering.                                  | Insert link |
+| [state-lifting](state-lifting)       | Editor's state is lifted higher up React tree to establish connection with a custom `textarea`. | Insert link |
+| [umd](umd)                           | Showcase of how `umd` build of `ckeditor4-react` package can be used.                          | Insert link |
```
Yeah, I forgot to run auto-formatting after recent change.

MMMalik commented on rollup.config.js:
```diff
@@ -0,0 +1,123 @@
+			format: 'cjs',
+			file: 'dist/index.cjs.js'
+		},
+		plugins: [ typescript(), cleanupPlugin() ]
```
After some considerations, I think it's better to use default `es5` target in both cases. It's not fair for developers targeting older browsers to transpile our library. For example, if they used webpack's `babel-loader` they would need to hack around `exclude: /node_modules/` to get our library transpiled. I will also get rid of `Object.entries` and `Array.prototype.includes` so that polyfills are not required for them.

On a side note, including polyfills in the library is a different story. I'd rather not do that. I will list required IE11 polyfills in the README though (it should be just `promise` I believe).

MMMalik commented:
Thanks for detailed review!

> I'd slightly change how test:e2e script works. Currently it does not exit with code 1 when the initial connection to BS can't be made. It should exit early in this case, because any subsequent tests will fail.

My intent was to re-try Nighwatch script due to following issues:
- connectivity issues with BS; I observed occasional problems with establishing connection
- false negative results on some browsers; This is pretty random and seems to be happening in few % of test cases

That being said, I've added a [check](https://github.com/ckeditor/ckeditor4-react/pull/199/files#diff-f9d0aca2e32dff9bca58e2a7a39135a698a7044e881a70b4040f0e96d11fcea5R39) at the beginning of e2e-runner so that it exits early on missing environment variables.

> All E2E tests passed, but the whole process didn't end with 0 exit code. It failed with the following error:

It looks like the way I handled closing of `express` app in `ssr` sample was very naive. I've fixed that by leveraging `tree-kill` package. It should take care of killing child processes.

> Something's wrong with setting inital data.

Good catch. Indeed, it looks like simply calling `setData` was not enough. Now I'm passing [config](https://github.com/ckeditor/ckeditor4-react/pull/199/files#diff-65e8aec743072216a0c4318cd15871fe1cac3f8288757c31933c97c0928ca391R69) to `setData` call so that undo stack is cleared.

> In component sample the behaviour of undo is weird.

It looks like passing config to `setData` fixed the `undo` stack problem too. Regardless, this one is still troubling me. My assumption was that one can re-use same DOM element for new instances of editor. Are there any known problems with this approach? If so, I will have to re-create new DOM element anytime `config` or `type` prop changes.

On a side-note, I'm wondering if there is a better approach to handle `initData`. Something like this maybe? WDYT?

```js
const CKEditor = ({ initData, ...otherProps }) => {
  // ...core logic

  return (
    <div
	id={name ?? undefined}
	ref={setElement}
	style={style ?? undefined}
    >
      {initData}
    </div>
  )
}
```

> Please update the list of used libs inside LICENSE.md file.

I've added `tslib` as parts of that library is bundled via `@rollup/plugin-typescript`.

Comandeer commented:
> On a side-note, I'm wondering if there is a better approach to handle `initData`. Something like this maybe? WDYT?

Looks like a really good idea :+1:

Comandeer reviewed (changes requested):
Looks good. Only some small changes.

Not sure if it's connected with how samples work or it's some more generic bug, but:

1. Open `hook` or `component` samples.
2. Switch editor to read-only mode and switch from classic to inline

SCAYT button is active, but clicking does nothing.

Probably it can be connected with the fact that the new editor is bound to the same element as the previous one.

The `ssr` sample is the only one working on port 3000 instead of 8080 as other samples. Maybe we can change it to 8080 for consistency with other samples?

I'm also wondering if the current code is able to handle changes proposed in ckeditor/ckeditor4#4601 (so reattaching the editor to the DOM)? These changes are intended to be included in 4.17.0, so not that distant future. It would be nice if we didn't have to do much refactor due to that.

MMMalik commented on src/events.ts:
```diff
@@ -0,0 +1,138 @@
+ *
+ * See: https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html
+ */
+export const events = [
```
Agreed, I will make that extendable by using generic types.

MMMalik commented on src/registerEditorEventHandler.ts:
```diff
@@ -0,0 +1,73 @@
+		};
+	}
+
+	if ( editor ) {
```
Yeah, that makes sense.

MMMalik commented:
Thanks @Comandeer!

I have applied the following changes:
1. Changed `ssr` default port.
2. Changed the way how `initData` works. Now it's just a child of root element. In order to hide the content before editor is initialized, I've added a small styling trick.
3. Memoized `config` and `type` props. See explanation below.
4. Added a possibility to listen to custom events. This is done via generic interface (both in `CKEditor` component and `useCKEditor` hook). In order to achieve that I had to refactor types related to event handling. There is now a new verbose `CKEditorEventHandlerProp` interface instead of dynamic [template literals](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html). This has one advantageous side-effect though: Now we support TypeScript as low as version 3.5. I've added new test cases and adjusted event samples.

Regarding your other comments:

> SCAYT button is active, but clicking does nothing.
> Probably it can be connected with the fact that the new editor is bound to the same element as the previous one.

This confirms what I suspected previously: One cannot re-use DOM element for creating new editor instances. Therefore I have memoized `config` and `type` props. Now re-mounting of `CKEditor` component must be done in userland via keyed component (pass unique `key` prop). This behavior is now quite similar to what we have in v1. I have refactored samples accordingly.

> I'm also wondering if the current code is able to handle changes proposed in ckeditor/ckeditor4#4601 (so reattaching the editor to the DOM)?

If I understand correctly, this should fix behavior of editor losing iframe context? If so, then in case of React integration this will fix the issue of rendering list of `CKEditor` components. I believe it should not affect React integration as long as editor handles that under the hood. It would be great though if I could test this new feature before its release. Is it possible to access `beta` version via CDN?

Comandeer commented on src/CKEditor.tsx:
```diff
@@ -0,0 +1,209 @@
+	/**
+	 * `readOnly` prop takes precedence over `config.readOnly`.
+	 */
+	if ( config && typeof readOnly === 'boolean' ) {
+		config.readOnly = readOnly;
+	}
```
And shouldn't it be also bound the other way? If `readOnly` is undefined but `config.readOnly` is set, then set `readOnly` to `config.readOnly` value.

MMMalik replied:
> What if config is not an object for some reason? In such a case editor won't become read only, which seems like a bug from the user's perspective.

Indeed, `config` can be `undefined`. In such case, if `readOnly` is set to true, editor will be set to `readOnly` mode later on in `useEffect` [here](https://github.com/ckeditor/ckeditor4-react/blob/48e916c5783a1edd94d30c0d04caf30d475f13ba/src/CKEditor.tsx#L116). I assumed that end-user won't notice a difference but it's probably better to set a default value for `config` like you suggested 👍.

> And shouldn't it be also bound the other way? If readOnly is undefined but config.readOnly is set, then set readOnly to config.readOnly value.

I'm not sure about this. I can imagine that `config.readOnly` is set to true and later on they want to pass `readOnly` as `false` during app runtime to opt-out from read-only mode. Isn't it how v1 works?

Comandeer replied:
> I'm not sure about this. I can imagine that `config.readOnly` is set to true and later on they want to pass `readOnly` as `false` during app runtime to opt-out from read-only mode. Isn't it how v1 works?

I meant only at the start. If `config.readOnly` is set, then set `readOnly`. It should allow modifying it later.

MMMalik replied:
> I meant only at the start. If config.readOnly is set, then set readOnly. It should allow modifying it later.

We could do that but I can't see how that would improve the component 🤔.

Comandeer commented on src/CKEditor.tsx:
```diff
@@ -0,0 +1,209 @@
+	/**
+	 * Sets and updates styles.
+	 */
+	useEffect( () => {
+		const canSetStyles =
+			type !== 'inline' &&
+			editor &&
+			( status === 'loaded' || status === 'ready' );
+
+		if ( style && canSetStyles ) {
+			editor.container.setStyles( style );
+		}
+
+		return () => {
+			if ( style && canSetStyles ) {
+				Object.keys( style )
+					.map( camelToKebab )
+					.forEach( styleName => {
+						editor.container.removeStyle( styleName );
+					} );
+			}
+		};
+	}, [ editor, status, style, type ] );
```
It seems that the component allows adding styles only to classic editors, but in tests and samples I can see that styles are also added to inline editors.

MMMalik replied:
Styling for inline editor [is set on root `div` element](https://github.com/ckeditor/ckeditor4-react/blob/48e916c5783a1edd94d30c0d04caf30d475f13ba/src/CKEditor.tsx#L123) via regular `style` prop. My understanding is that in inline that root `div` is used by the editor while in classic mode a new root for editor is created. Therefore, for classic mode we have to imperatively change styles.

Comandeer commented on src/CKEditor.tsx:
```diff
@@ -0,0 +1,209 @@
+	return (
+		<div id={name ?? undefined} ref={setElement} style={style ?? undefined}>
+			{initData ? (
+				<div style={{ display: 'none' }}>{initData}</div>
```
Doing so you in fact insert invisible `div` into editor's content, which can have weird consequences if connected with certain editor's config, e.g.

```javascript
<CKEditor debug={true} initData="Hello world!" config={{ extraAllowedContent: 'div{*}' }} />
```

will result in "empty" editor (the one containing hidden `div`).

In fact you should just insert raw `initData` into the editor's element.

Comandeer replied:
There is another issue with hidden `div`: it changes the way in which SSR works (it does not longer render the visible text).

MMMalik replied:
Yeah, that was a hacky way to hide initial content before editor is ready. I need to find a better way. Do you have any hints?

Comandeer replied:
I'm not sure if it's even an issue, TBH. CKE4 works that way from the beginning: the content is just visible before editor's initialisation.

MMMalik replied:
I've given it one more try. I've added `getStyle` util that returns `style` for the root element based on editor `type` and `status`. It seems to be working fine.

Comandeer reviewed (changes requested):
> 2. Changed the way how `initData` works. Now it's just a child of root element. In order to hide the content before editor is initialized, I've added a small styling trick.

It seems to work, but please see inline comment about it.

> Is it possible to access `beta` version via CDN?

Maybe not beta, but there is [nightly](https://nightly.ckeditor.com/).

---

Looks pretty well, just several small issues. Please also check if code works with the newest nightly (as it includes reattach feature).

MMMalik commented:
Thanks @Comandeer.

I've added following changes:
- initialized `config` in `CKEditor` component
- bumped editor version to 4.16.1
- fixed root styling for `initData`

> It seems to work, but please see inline comment about it.

I've given it one more try. I still think that setting initial data via `children` is slightly more elegant solution than using `editor.setData` imperatively. I will use the latter way as a last resort though.

Regarding nightly build. As I mentioned on Slack, I did some testing with both versions of React integration and it looks like no config adjustments are required. Nightly build just works out of the box. And it fixes the issue with re-ordering classic editors 🎉.

Comandeer reviewed (approved):
LGTM! 🎉

Could you only rebase it onto `major` and merge it?
